pkg/translate/translate.go

```diff
@@ -406,6 +445,55 @@ func (t *Translator) OverlayK8sSettings(yml string, icp *v1alpha2.IstioControlPl
 	return objects.YAMLManifest()
 }
 
+// mergeEnv apply deep merge overlay env settings on the base env.
+func mergeEnv(baseEnv, overlayEnv string) (string, error) {
+	if overlayEnv == "" {
+		return baseEnv, nil
+	}
+	baseEnvJSON, err := yaml.YAMLToJSON([]byte(baseEnv))
+	if err != nil {
+		return "", fmt.Errorf("yamlToJSON error in base env: %s\n%s", err, baseEnv)
+	}
+	overlayEnvJSON, err := yaml.YAMLToJSON([]byte(overlayEnv))
+	if err != nil {
+		return "", fmt.Errorf("yamlToJSON error in overlay env: %s\n%s", err, overlayEnv)
+	}
+
+	var baseEnvObj, overlayEnvObj []*v1.EnvVar
+	err = json.Unmarshal(baseEnvJSON, &baseEnvObj)
+	if err != nil {
+		return "", err
+	}
+	err = json.Unmarshal(overlayEnvJSON, &overlayEnvObj)
+	if err != nil {
+		return "", err
+	}
+
+	dstEnvObj := make([]*v1.EnvVar, len(baseEnvObj))
+	copy(dstEnvObj, baseEnvObj)
+	for i, v2 := range overlayEnvObj {
+		exists := false
+		for j, v1 := range baseEnvObj {
+			if v1.Name == v2.Name {
+				// if the env item with name exists in base, then override base.
+				exists = true
+				dstEnvObj[j] = overlayEnvObj[i]
+				break
+			}
+		}
+		if !exists {
+			// if the env item with name doesn't exist in base, then append to base.
+			dstEnvObj = append(dstEnvObj, overlayEnvObj[i])
+		}
+	}
+
+	b, err := json.Marshal(dstEnvObj)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
```

Comment:
It will simplify the logic considerably if you think of the problem in the following way: the root issue here is that overlayK8s below is doing an overwrite using WriteNode, whereas we should be doing a merge for every node. You could either create a MergeNode wrapper similar to WriteNode that internally calls the traversal code and does a merge when the node is found, or you can read the node here, merge with the new value and WriteNode the result. 
In either case you can use overlayTrees or OverlayYAML to do the merge.

Reply:
btw. feel free to ping me tonight on slack is any of this doesn't make sense or you need some more pointers.

Reply:
@ostromart Have tried the overlayTrees and OverlayYAML, it can't meet our requirements. The reason is that we currently use the general json/yaml merge tool, eg. https://github.com/evanphx/json-patch
For `env`, we have original snippet:

```
env:
- name: n1
  value: v1
```

and override snippet:
```
env:
- name: n1
  value: v2
```

the general json/yaml merge will just reproduce:
```
env:
- name: n1
  value: v1
- name: n1
  value: v2
```

we need semantic merge, because what we really want is:
```
env:
- name: n1
  value: v2
```

Reply:
Creating a MergeNode wrapper similar to WriteNode make sense to me.

Reply:
That's a good point, we would need extra work to implement the correct merge semantics. However, I think if this code is added to a MergeNode function I think it would be a reasonable default behavior for that function and be generally useful. 
Thanks for tackling this tricky problem btw. :-)

Reply:
Actually I'm thinking to add a general method, eg, `K8sJsonMerge` that accepts base and overlay json, also somewhat about **the keys of array object**, but it's tricky, because I can't figure out how to pass the keys of array object.
Wdyt, does it make sense?

Reply:
Yes, it's tricky! I can suggest the following approach: recurse through the base tree A and for each node:
- if node doesn't exist in B, skip.
- if the node is a leaf-list, find the item that has the most keys in common and overlay B on that item. if no item matches on any key, append to the list.
- if the node is non-leaf list, find an item that matches B item exactly and continue to traverse.
Otherwise if that proves to be too much work, you could just do a JSON strategic merge on the trees. We probably don't have to handle such tricky cases as above, we can probably assume that lists are non-overlapping, otherwise it's a user error. For example, if the same env var has two different values in different places, we can treat this as a bad configuration. If it is duplicated, I think strategic merge will handle it.

Reply:
The JSON merge package https://github.com/evanphx/json-patch we're using now doesn't handle slice/array well, it just replace all items with new items, instead of doing a JSON strategic merge, check this example: https://play.golang.org/p/woFuyU1vWKU

Reply:
Switched to strategic merge.